File: implementations/c/binary_search/binary_search_mid.c
#include <stdio.h>
#include <stdint.h>
#include <inttypes.h>

#define N 10000

static int64_t arr[N];

static void setup(void) {
    for (int i = 0; i < N; i++) {
        arr[i] = (int64_t)i * 2;
    }
}

static int binary_search(void) {
    int64_t target = (int64_t)(N - 1) * 2;
    int lo = 0, hi = N - 1;
    while (lo <= hi) {
        int mid = lo + (hi - lo) / 2;
        if      (arr[mid] == target) return mid;
        else if (arr[mid] < target)  lo = mid + 1;
        else                         hi = mid - 1;
    }
    return -1;
}

int main(void) {
    setup();

    printf("ready\n");
    fflush(stdout);

    char buf[64];
    while (fgets(buf, sizeof(buf), stdin)) {
        printf("%d\n", binary_search());
        fflush(stdout);
    }

    return 0;
}

File: implementations/java/binary_search/BinarySearch_mid.java
package binary_search;

import java.io.*;

public class BinarySearch_Mid {
    static final int N = 10000;
    static long[] arr = new long[N];

    static void setup() {
        for (int i = 0; i < N; i++) {
            arr[i] = (long)i * 2;
        }
    }

    static int binarySearch() {
        long target = (long)(N - 1) * 2;
        int lo = 0, hi = N - 1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (arr[mid] == target)      return mid;
            else if (arr[mid] < target)  lo = mid + 1;
            else                         hi = mid - 1;
        }
        return -1;
    }

    public static void main(String[] args) throws IOException {
        setup();

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        System.out.println("ready");
        System.out.flush();

        String line;
        while ((line = br.readLine()) != null) {
            System.out.println(binarySearch());
            System.out.flush();
        }
    }
}